src/hpke.rs

```diff
@@ -0,0 +1,610 @@
+}
+
+fn open<T: Aead>(
+    enc: x25519::PublicKey,
+    skr: x25519::SecretKey,
+    info: &[u8],
+    aad: &[u8],
+    ciphertext: &mut [u8],
+    tag: [u8; TAG_LEN],
+) -> Result<(), aead::Error> {
+    let ctx = setup_base_r::<T>(enc, skr, info);
+    ctx.open_in_place_detached::<T>(aad, ciphertext, tag)
+}
+
+/// Open with X25519-HKDF-SHA256-ChaCha20Poly1305 suite
+fn open8(
+    enc: x25519::PublicKey,
+    skr: x25519::SecretKey,
+    info: &[u8],
+    aad: &[u8],
+    ciphertext: &mut [u8],
+    tag: [u8; TAG_LEN],
+) -> Result<(), aead::Error> {
+    open::<ChaCha8Poly1305>(enc, skr, info, aad, ciphertext, tag)
+}
+
+fn load_public_key(
+    key_id: &KeyId,
+    keystore: &mut impl Keystore,
+) -> Result<x25519::PublicKey, trussed::Error> {
+    let public_bytes: [u8; 32] = keystore
+        .load_key(key::Secrecy::Public, Some(key::Kind::X255), key_id)?
+        .material
+        .as_slice()
+        .try_into()
+        .map_err(|_| trussed::Error::InternalError)?;
+    let public_key = x25519::PublicKey::from(public_bytes);
+    Ok(public_key)
+}
+
+fn load_secret_key(
+    key_id: &KeyId,
+    keystore: &mut impl Keystore,
+) -> Result<x25519::SecretKey, trussed::Error> {
+    let secret_bytes: [u8; 32] = keystore
+        .load_key(key::Secrecy::Secret, Some(key::Kind::X255), key_id)?
+        .material
+        .as_slice()
+        .try_into()
+        .map_err(|_| trussed::Error::InternalError)?;
+    let secret_key = x25519::SecretKey::from_seed(&secret_bytes);
+    Ok(secret_key)
+}
+
+impl ExtensionImpl<HpkeExtension> for StagingBackend {
+    fn extension_request<P: trussed::Platform>(
+        &mut self,
+        core_ctx: &mut trussed::types::CoreContext,
+        _backend_ctx: &mut Self::Context,
+        request: &<HpkeExtension as trussed::serde_extensions::Extension>::Request,
+        resources: &mut trussed::service::ServiceResources<P>,
+    ) -> Result<<HpkeExtension as trussed::serde_extensions::Extension>::Reply, trussed::Error>
+    {
+        let filestore = &mut resources.filestore(core_ctx.path.clone());
+        let keystore = &mut resources.keystore(core_ctx.path.clone())?;
+
+        match request {
+            HpkeRequest::Seal(req) => {
+                let mut pt = req.plaintext.clone();
+                let public_key = load_public_key(&req.key, keystore)?;
+                let (pk, tag) = seal8(public_key, &req.info, &req.aad, &mut pt, keystore.rng());
+                let enc = keystore.store_key(
+                    req.enc_location,
+                    key::Secrecy::Public,
+                    key::Kind::X255,
+                    &pk.to_bytes(),
+                )?;
+                Ok(HpkeSealReply {
+                    enc,
+                    ciphertext: pt,
+                    tag: tag.into(),
+                }
+                .into())
+            }
+            HpkeRequest::SealKey(req) => {
+                // TODO: need to check both secret and private keys
```

Comment:
I don’t get this comment.  Can you elaborate?

Reply:
One might want to seal a public key. Not used for  now.

Reply:
I think I’m still missing something.  Aren’t “secret” and “private” synonyms in Trussed?

Reply:
Oh, it's a typo.

Reply:
Makes sense now. :D  The same comment is also on line 374.